plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -184,23 +182,89 @@ typeMismatchError uniExp uniAct = fromString $ concat
+-- Equivalent to @ExceptT KnownTypeError Emitter@, except optimized in two ways:
+--
+-- 1. everything is strict
+-- 2. has the 'MakeKnownSuccess' constructor that is used for returning a value with no logs
+--    attached, which is the most common case for us, so it helps a lot not to construct and
+--    deconstruct a redundant tuple
+--
+-- Moving from @ExceptT KnownTypeError Emitter@ to this data type gave us a speedup of 8% of total
+-- evaluation time.
+data MakeKnownM a
+    = MakeKnownFailure !(DList Text) !KnownTypeError
+    | MakeKnownSuccess !a
+    | MakeKnownSuccessWithLogs !(DList Text) !a
+
+-- | Prepend logs to a 'MakeKnownM' computation and 'fmap' it.
+fmapWithLogs :: DList Text -> (a -> b) -> MakeKnownM a -> MakeKnownM b
+fmapWithLogs logs1 f = \case
+    MakeKnownFailure logs2 err       -> MakeKnownFailure (logs1 <> logs2) err
+    MakeKnownSuccess x               -> MakeKnownSuccessWithLogs logs1 (f x)
+    MakeKnownSuccessWithLogs logs2 x -> MakeKnownSuccessWithLogs (logs1 <> logs2) (f x)
+{-# INLINE fmapWithLogs #-}
+
+-- | Prepend logs to a 'MakeKnownM' computation.
+withLogs :: DList Text -> MakeKnownM a -> MakeKnownM a
+withLogs logs1 = \case
+    MakeKnownFailure logs2 err       -> MakeKnownFailure (logs1 <> logs2) err
+    MakeKnownSuccess x               -> MakeKnownSuccessWithLogs logs1 x
+    MakeKnownSuccessWithLogs logs2 x -> MakeKnownSuccessWithLogs (logs1 <> logs2) x
+{-# INLINE withLogs #-}
+
+instance Functor MakeKnownM where
+    -- Written out explicitly, because for some inexplicable reason GHC fails to inline
+    -- @fmapWithLogs mempty@ despite the pragma.
+    fmap _ (MakeKnownFailure logs err)       = MakeKnownFailure logs err
+    fmap f (MakeKnownSuccess x)              = MakeKnownSuccess (f x)
+    fmap f (MakeKnownSuccessWithLogs logs x) = MakeKnownSuccessWithLogs logs (f x)
+    {-# INLINE fmap #-}
+
+    -- Written out explicitly just in case (see @fmap@ above for what the case might be).
+    _ <$ MakeKnownFailure logs err       = MakeKnownFailure logs err
+    x <$ MakeKnownSuccess _              = MakeKnownSuccess x
+    x <$ MakeKnownSuccessWithLogs logs _ = MakeKnownSuccessWithLogs logs x
+    {-# INLINE (<$) #-}
+
+instance Applicative MakeKnownM where
+    pure = MakeKnownSuccess
+    {-# INLINE pure #-}
+
+    MakeKnownFailure logs err       <*> _ = MakeKnownFailure logs err
+    MakeKnownSuccess f              <*> a = fmap f a
+    MakeKnownSuccessWithLogs logs f <*> a = fmapWithLogs logs f a
+    {-# INLINE (<*>) #-}
+
+    MakeKnownFailure logs err       *> _ = MakeKnownFailure logs err
+    MakeKnownSuccess _              *> a = a
+    MakeKnownSuccessWithLogs logs _ *> a = withLogs logs a
+    {-# INLINE (*>) #-}
+
+instance Monad MakeKnownM where
+    MakeKnownFailure logs err       >>= _ = MakeKnownFailure logs err
+    MakeKnownSuccess x              >>= f = f x
+    MakeKnownSuccessWithLogs logs x >>= f = withLogs logs $ f x
```

Comment:
This could lead to a linear number of `withLogs` (i.e., `withLogs logs1 $ withLogs logs2 $ ... $ MakeKnownSuccessWithLogs logs_n x`) before starting to merge the logs, but I guess it doesn't really matter since logs are DLists?

Reply:
Yep!

Reply:
> This could lead to a linear number of `withLogs`

I'm nitpicking, but in the vast majority of cases `withLogs` is going to be computed (because it's not easy to accidentally force `f` without forcing `withLogs`) and only `logs` is going to be retained. That last part is the classic strict `Writer` [leak](https://mail.haskell.org/pipermail/libraries/2013-March/019528.html) indeed.

But yes, we don't care about retaining `logs` at every step until the very end given that their reduced-to-WHNF version `logs1 . logs2` is no shorter than retaining them separately. That would be a problem for `x1 + x2 + x3 + ...`, because evaluating that does reduce the space taken, but in our case it's fine.

Also we don't log much within the definition of a builtin, I think there are only two builtins that do logging, so we are not going to returns lots of logs from builtins anyway.